### DISCUSSION TITLE
submitting a link to a discord bot and having it put it into a txt file

### DISCUSSION THREAD

**Discussion by peppergj** [Rapptz/discord.py#9974] [Q&A]
so i'm currently making a discord bot which helps out with an album exchange which my friend group uses.
and so basically, i can't find anything that can help out with being able to submit spotify/album links to the bot, so it puts it into a txt file, then when i draw the submitted albums, it randomly auto-selects one.
can anyone help out or point me into the right direction?

**Comment by nihilne** [comment]
It really depends on how you want to implement this. When I say "it depends", it's all about the needs of your app. If you need a robust way to store data, then you have to go with a database implementation (e.g. MySQL), but if you just want to store plaintext data and then parse it (in your case, links/URLs), a .json file (or even a .txt file as you mentioned) will be enough.

My suggestion is, instead of using a text file (to avoid formatting and parsing problems) you should use a JSON file to store your links. You can then easily read the file when you want to randomly select a link. There are many examples, both in StackOverflow and the discord.py Discord server (and on the internet in general).

**Reply by peppergj** [reply]
thanks for your reply! 
i was more wondering the kind of code i would need to learn/write to have the bot take link submissions.
do you know anything about this?

**Reply by Drakan1** [reply]
That's relatively easy. So you can do it as an argument from a standard function or use a [wait_for](https://discordpy.readthedocs.io/en/stable/ext/commands/api.html?highlight=wait#discord.ext.commands.Bot.wait_for)

Something like:
```py
class Music:
 def __init__(self, name, link, requested_by etc):
....


class MusicCommand (commands.cog):
...
@commands.command()
async def submission(self, ctx, link):
 # some form of check here
    self.links.append(Music(ctx.author.name, link etc)) # can test for whether the link is valid

```

**Comment by felipeyx** [comment]
You don't need the Spotify API just to store submitted links and randomly choose one.

For a small bot, you can append each submission to a text file:

import random
from discord.ext import commands

@bot.command()
async def submit(ctx, link: str):
    if not link.startswith(
        ("https://open.spotify.com/", "https://spotify.link/")
    ):
        await ctx.send("Please submit a Spotify link.")
        return

    with open("albums.txt", "a", encoding="utf-8") as f:
        f.write(link + "\n")

    await ctx.send("Album submitted.")

Then randomly draw one:

@bot.command()
async def draw(ctx):
    with open("albums.txt", "r", encoding="utf-8") as f:
        albums = [
            line.strip()
            for line in f
            if line.strip()
        ]

    if not albums:
        await ctx.send("No albums have been submitted.")
        return

    await ctx.send(random.choice(albums))

For a friend-group bot this is enough. If you later need to prevent duplicate submissions, track who submitted each album, or remove entries after drawing, SQLite would be a better next step than a text file.
